File: src/main/java/com/quid/kopring/service/book/BookService.java
package com.quid.kopring.service.book;

import com.quid.kopring.book.Book;
import com.quid.kopring.book.repository.BookJpaRepository;
import com.quid.kopring.dto.book.request.BookLoanRequest;
import com.quid.kopring.dto.book.request.BookRequest;
import com.quid.kopring.dto.book.request.BookReturnRequest;
import com.quid.kopring.user.User;
import com.quid.kopring.user.repository.UserJpaRepository;
import com.quid.kopring.userLoanHistory.repository.UserLoanHistoryJpaRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class BookService {

    private final BookJpaRepository bookRepository;
    private final UserJpaRepository userRepository;
    private final UserLoanHistoryJpaRepository userLoanHistoryRepository;

    @Transactional
    public void saveBook(BookRequest request) {
        Book newBook = new Book(null, request.getName());
        bookRepository.save(newBook);
    }

    @Transactional
    public void loanBook(BookLoanRequest request) {
        Book book = bookRepository.findByName(request.getBookName())
            .orElseThrow(IllegalArgumentException::new);
        if (userLoanHistoryRepository.findByBookNameAndIsReturn(request.getBookName(), false)
            != null) {
            throw new IllegalArgumentException("진작 대출되어 있는 책입니다");
        }

        User user = userRepository.findByName(request.getUserName())
            .orElseThrow(IllegalArgumentException::new);
        user.loanBook(book);
    }

    @Transactional
    public void returnBook(BookReturnRequest request) {
        User user = userRepository.findByName(request.getUserName())
            .orElseThrow(IllegalArgumentException::new);
        user.returnBook(request.getBookName());
    }

}

File: src/main/kotlin/com/quid/kopring/book/service/BookService.kt
package com.quid.kopring.book.service

import com.quid.kopring.book.Book
import com.quid.kopring.book.model.request.BookCreateRequest
import com.quid.kopring.book.model.request.BookLoanRequest
import com.quid.kopring.book.model.request.BookReturnRequest
import com.quid.kopring.book.model.request.BookUpdateRequest
import com.quid.kopring.book.model.response.BookStat
import com.quid.kopring.book.repository.BookRepository
import com.quid.kopring.user.repository.UserJpaRepository
import com.quid.kopring.userLoanHistory.repository.UserLoanHistoryJpaRepository
import com.quid.kopring.userLoanHistory.type.UserLoanStatus
import com.quid.kopring.util.fail
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class BookService(
    private val bookRepository: BookRepository,
    private val userJpaRepository: UserJpaRepository,
    private val userLoanHistoryRepository: UserLoanHistoryJpaRepository
) {
    @Transactional
    fun saveBook(request: BookCreateRequest) {
        Book(name = request.name, type = request.type).also { bookRepository.save(it) }
    }

    @Transactional
    fun loanBook(request: BookLoanRequest) {
        val book = bookRepository.findByName(request.bookName) ?: fail("Book not found")

        if (userLoanHistoryRepository.findByBookNameAndStatus(
                request.bookName, UserLoanStatus.LOANED
            ) != null
        ) fail("Book is already loaned")

        val user = userJpaRepository.findByName(request.userName) ?: fail("User not found")
        user.loanBook(book)
    }

    @Transactional
    fun returnBook(request: BookReturnRequest) {
        userJpaRepository.findByName(request.userName)?.returnBook(request.bookName)
            ?: fail("User not found")
    }

    @Transactional(readOnly = true)
    fun getBooks(): List<Book> {
        return bookRepository.findAll()
    }

    @Transactional(readOnly = true)
    fun getCountLoanBook(): Int {
        return userLoanHistoryRepository.findByStatus(UserLoanStatus.LOANED).size
    }

    @Transactional(readOnly = true)
    fun getStat(): List<BookStat> {
        return bookRepository.getBookStat()
    }

    @Transactional
    fun updateBook(id: Long, request: BookUpdateRequest) {
        bookRepository.findById(id)?.update(request) ?: fail("Book not found")
    }
}